packages/block-editor/src/components/block-list/use-block-props/use-block-refs.js

```diff
@@ -37,30 +36,35 @@ export function useBlockRefProvider( clientId ) {
 }
 
 /**
- * Gets a ref pointing to the current block element. Continues to return the same
- * stable ref object even if the `clientId` argument changes. This hook is not
- * reactive, i.e., it won't trigger a rerender of the calling component if the
- * ref value changes. For reactive use cases there is the `useBlockElement` hook.
+ * Tracks the DOM element for the block identified by `clientId` and assigns it to the `ref`
+ * whenever it changes.
  *
- * @param {string} clientId The client ID to get a ref for.
- *
- * @return {RefObject} A ref containing the element.
+ * @param {string} clientId The client ID to track.
+ * @param {Ref}    ref      The ref object/callback to assign to.
  */
-function useBlockRef( clientId ) {
+export function useBlockElementRef( clientId, ref ) {
 	const { refsMap } = useContext( BlockRefs );
-	const latestClientId = useRef();
-	latestClientId.current = clientId;
-
-	// Always return an object, even if no ref exists for a given client ID, so
-	// that `current` works at a later point.
-	return useMemo(
-		() => ( {
-			get current() {
-				return refsMap.get( latestClientId.current ) ?? null;
-			},
-		} ),
-		[ refsMap ]
+	const setRef = useCallback(
+		( el ) => {
+			if ( typeof ref === 'function' ) {
+				ref( el );
+			} else if ( ref ) {
+				ref.current = el;
+			}
+		},
+		[ ref ]
 	);
+
+	useLayoutEffect( () => {
+		setRef( refsMap.get( clientId ) );
+		const unsubscribe = refsMap.subscribe( clientId, () =>
+			setRef( refsMap.get( clientId ) )
+		);
+		return () => {
+			unsubscribe();
+			setRef( null );
+		};
+	}, [ refsMap, clientId, setRef ] );
```

Comment:
There's one subtle difference with the previous implementation: the element was available immediately on render. That's probably not needed, but just mentioning it here in case it was relied upon before.

Reply:
This is not true since #63565. On initial render, the return value is `null`. Only in an effect scheduled by this render is it set to the actual element.

When the element was available immediately on render, it was returning references to stale elements.

Reply:
Oh ok! Thanks for clarifying